packages/ws-client/src/data-source.ts

```diff
@@ -0,0 +1,127 @@
+import type { DataSource, JSONValue, Store, TemplateEvent } from '@react-native-templates/core'
+import ReconnectingWebSocket from 'reconnecting-websocket'
+
+export type SubscribeWebsocketMessage = {
+  $: 'sub'
+  keys: string[]
+}
+
+export type UnsubscribeWebsocketMessage = {
+  $: 'unsub'
+  keys: string[]
+}
+
+export type EventWebsocketMessage = {
+  $: 'evt'
+  event: TemplateEvent
+}
+
+export type UpdateWebsocketMessage = {
+  $: 'up'
+  key: string
+  val: JSONValue
+}
+
+export type ClientWebsocketMessage =
+  | SubscribeWebsocketMessage
+  | UnsubscribeWebsocketMessage
+  | EventWebsocketMessage
+
+export type ServerWebsocketMessage = UpdateWebsocketMessage
+
+type Handler = () => void
+
+export function createWSDataSource(
+  wsUrl: string,
+  interceptEvent?: (event: TemplateEvent) => boolean
+): DataSource {
+  const rws = new ReconnectingWebSocket(wsUrl)
+  function send(payload: ClientWebsocketMessage) {
+    rws.send(JSON.stringify(payload))
+  }
+
+  const subscriptions = new Map<string, Set<Handler>>()
+  const cache = new Map<string, any>()
+
+  rws.addEventListener('open', () => {
+    const keys = [...subscriptions.entries()]
+      .filter(([, handlers]) => handlers.size > 0)
+      .map(([key]) => key)
+    if (keys.length === 0) {
+      return
+    }
+    send({
+      $: 'sub',
+      keys,
+    })
+  })
+
+  rws.onmessage = (eventData) => {
+    const event = JSON.parse(eventData.data) as ServerWebsocketMessage
```

Comment:
we should validate that the message is correct, using zod or something. this is technically not safe

Reply:
Yeah, we need to add runtime type checking to many places, including this. I will handle that in a follow-up PR.